keeper-gradle-plugin/src/test/kotlin/com/slack/keeper/KeeperFunctionalTest.kt

```diff
@@ -0,0 +1,365 @@
+/*
+ * Copyright (C) 2020 Slack Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.slack.keeper
+
+import com.google.common.truth.Truth.assertThat
+import com.squareup.javapoet.ClassName
+import org.gradle.testkit.runner.BuildResult
+import org.gradle.testkit.runner.GradleRunner
+import org.gradle.testkit.runner.TaskOutcome
+import org.intellij.lang.annotations.Language
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.TemporaryFolder
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+import org.junit.runners.Parameterized.Parameters
+import java.io.File
+import java.util.zip.ZipFile
+import javax.lang.model.element.Modifier.STATIC
+import com.squareup.kotlinpoet.ClassName as KpClassName
+
+/**
+ * Testing gradle plugins is finicky. If you get errors when running from the IDE, try following
+ * these instructions: https://stackoverflow.com/a/44692954/3323598
+ *
+ * Alternatively - run `./gradlew pluginUnderTestMetadata` once from the command line to generate
+ * the metadata file and then try again from the IDE.
+ *
+ * To debug the Gradle plugin's code itself, uncomment the `.withDebug()` line in [runGradle]
+ * function. Normal debugging in the test code doesn't require this.
+ *
+ * ---
+ *
+ * The basic test project structure is roughly this:
+ *
+ * ```
+ * projectDir
+ *   - build.gradle
+ *   - proguardConfigOutput.pro
+ *   - src
+ *     - androidTest
+ *       - AndroidManifest.xml
+ *       - java/com/slack/keeper/example
+ *         - ApplicationUsedClass.java
+ *         - SampleApplication.java
+ *         - TestOnlyClass.java
+ *         - TestOnlyKotlinClass.kt
+ *         - UnusedClass.java
+ *     - main
+ *       - AndroidManifest.xml
+ *       - java/com/slack/keeper/example
+ *         - TestOnlyClassCaller.java
+ *         - TestOnlyKotlinClassCaller.kt
+ * ```
+ */
+@RunWith(Parameterized::class)
+class KeeperFunctionalTest(private val useR8: Boolean) {
+
+  companion object {
+    @JvmStatic
+    @Parameters(name = "useR8={0}")
+    fun data(): List<Array<*>> {
+      return listOf(
+          arrayOf(true),
+          arrayOf(false)
+      )
+    }
+  }
+
+  @Rule
+  @JvmField
```

Comment:
Is this required? What if you changed the first annotation to `@get:Rule`? I'm pretty sure I've used Rules in Kotlin tests without needing jvm field. Obviously not a big deal and I'm just bikeshedding.

Reply:
Side note, I follow a different approach, and instead specify a real directory inside `build/`. The advantage of this is that, when there are issues, I can manually debug by looking at the generated project structure directly. Disadvantage is you need to take care to manually delete the directory after each test to ensure they're hermetic. A `clean` would also delete them.

Reply:
re: the rule - either way works. I like making it a field rather than apply the rule on the getter

Interesting point about `build`. I've found jumping around the debugger mostly ok, and was worried about creating a lot of throwaway files, especially if the path to a build dir is somehow messed up during running and they get dumped somewhere else.